miniscript: Use Func and Expr when parsing keys, hashes, and locktimes

achow101 opened the pull request:
The miniscript parser currently only looks for the next `)` when parsing key, hash, and locktime expressions. This fails to parse when the expressions contain a nested expression. Currently, this is only possible with `musig()` inside of key expressions. However, this pattern can be generalized to handling hashes and locktimes, so I implemented those too.

Fixes #34076

achow101 committed:
miniscript: Using Func and Expr when parsing keys, hashes, and locktimes

Since pk(), pk_k(), pkh(), pk_h(), sha256(), ripemd160(), hash256(),
hash160(), after(), and older() all are single argument expressions that
are parsed immediately, we can use the Expr and Func parsing functions
to determine what the arguments of these expressions are, rather than
searching for the next closing parentheses.

This fixes an issue when pk(), pk_k(), pkh(), and pk_h() include a
musig() expression as Expr properly handles nested expressions.

achow101 committed:
test: Test for musig() in various miniscript expressions

scgbckbone reviewed (commented):
Import now works OK, but there is a problem with address generation. Test case:
```
self.test_success_case("time-locked musig",
                     "tr(musig($0,$1,$2)/<0;1>/*,{and_v(v:pk(musig($0,$1)/<0;1>/*),after(1)),{and_v(v:pk(musig($0,$2)/<0;1>/*),after(1)),and_v(v:pk(musig($1,$2)/<0;1>/*),after(1))}})",
                     scriptpath=True)
```

Particular case:
```
core_desc_object = [{'desc': 'tr(musig([0f056943/86h/1h/0h]tpubDCeEX49avtiXrBTv3JWTtco99Ka499jXdZHBRtm7va2gkMAui11ctZjqNAT9dLVNaEozt2C1kfTM88cnvZCXsWLJN2p4viGvsyGjtKVV7A1,[fdc7b3da/44h/1h/0h]tpubDDdbJUqUksL8pDjkh18Z6YC3kvBW7zufRKnawvvToRh3oTYVvuV3mNyEgav7VZf7JVMvSUxDxAc1MipVETVe6jmTzcaTNMRGf3Gh3eCQFsY,[44fbc2f9/44h/1h/0h]tpubDDMKx1DX6V3wfubHm2jSX11ToHMdYYwmtS3Sm3VJ5ky5efECGPa1kN3zN8hqdeEqrbqez33f8YHLARxUoEGkr2r68ViwXWAxhVcU8Tw1pSV)/<2;3>/*,{and_v(v:pk(musig([0f056943/86h/1h/0h]tpubDCeEX49avtiXrBTv3JWTtco99Ka499jXdZHBRtm7va2gkMAui11ctZjqNAT9dLVNaEozt2C1kfTM88cnvZCXsWLJN2p4viGvsyGjtKVV7A1,[44fbc2f9/44h/1h/0h]tpubDDMKx1DX6V3wfubHm2jSX11ToHMdYYwmtS3Sm3VJ5ky5efECGPa1kN3zN8hqdeEqrbqez33f8YHLARxUoEGkr2r68ViwXWAxhVcU8Tw1pSV)/<0;1>/*),older(10)),{and_v(v:pk(musig([0f056943/86h/1h/0h]tpubDCeEX49avtiXrBTv3JWTtco99Ka499jXdZHBRtm7va2gkMAui11ctZjqNAT9dLVNaEozt2C1kfTM88cnvZCXsWLJN2p4viGvsyGjtKVV7A1,[fdc7b3da/44h/1h/0h]tpubDDdbJUqUksL8pDjkh18Z6YC3kvBW7zufRKnawvvToRh3oTYVvuV3mNyEgav7VZf7JVMvSUxDxAc1MipVETVe6jmTzcaTNMRGf3Gh3eCQFsY)/<0;1>/*),older(10)),and_v(v:pk(musig([fdc7b3da/44h/1h/0h]tpubDDdbJUqUksL8pDjkh18Z6YC3kvBW7zufRKnawvvToRh3oTYVvuV3mNyEgav7VZf7JVMvSUxDxAc1MipVETVe6jmTzcaTNMRGf3Gh3eCQFsY,[44fbc2f9/44h/1h/0h]tpubDDMKx1DX6V3wfubHm2jSX11ToHMdYYwmtS3Sm3VJ5ky5efECGPa1kN3zN8hqdeEqrbqez33f8YHLARxUoEGkr2r68ViwXWAxhVcU8Tw1pSV)/<0;1>/*),older(10))}})#3xfc5d0t', 'active': True, 'range': [0, 100], 'timestamp': 'now'}]


ms = bitcoind.create_wallet(
    wallet_name=name, disable_private_keys=True,
    blank=True, passphrase=None, avoid_reuse=False, descriptors=True
)

res = ms.importdescriptors(core_desc_object)
# res = [{'success': True}]

addr = ms.getnewaddress("", "bech32m")

# just random other wallet with funds
x = bitcoind.supply_wallet.sendtoaddress(addr, 49)
# x = a166d63b05162525752a1c911454eb7e3e27a3f509b2efcd5d74069d98de4382

# mine
y =bitcoind.supply_wallet.generatetoaddress(1,bitcoind.supply_wallet.getnewaddress())
# y = ['1826120d4e15d6c880ec5633ca74912bbae3a28c749a502de15379ce39af61a2']

# must have unspent now - but we don't
unspent = ms.listunspent()
# unspent = []

# first external address is not even recognized as ours
ms.getaddressinfo(addr)
# {'address': 'bcrt1pwvnfa5ypmfvnmcguwllehxqreq9ljrvz58hgeqxpwy9xqfta7r6scptqp0', 
# 'scriptPubKey': '512073269ed081da593de11c77ff9b9803c80bf90d82a1ee8c80c1710a60257df0f5', 
# 'ismine': False, 'solvable': False, 'iswatchonly': False, 'isscript': True, 'iswitness': True, 
# 'witness_version': 1,
# 'witness_program': '73269ed081da593de11c77ff9b9803c80bf90d82a1ee8c80c1710a60257df0f5', 
# 'ischange': False, 'labels': ['']}

# generating next address (index 1) works, and is recognized as ours
next_addr = ms.getnewaddress("", "bech32m")
bitcoind.supply_wallet.sendtoaddress(next_addr, 10)
bitcoind.supply_wallet.generatetoaddress(1, bitcoind.supply_wallet.getnewaddress())
ms.listunspent()
# [{'txid': 'bd893fdd25dced57a193cc8b5fe15bc8b7309744c5156c7932712e4e34f7141f', 'vout': 1, 'address': 'bcrt1pgkt8zf54dgkpf66zt56etr8ksctwczn453ec4m2gqydknstuk7mqg0rrlu', 'label': '', 'scriptPubKey': '512045967126956a2c14eb425d35958cf68616ec0a75a4738aed48011b69c17cb7b6', 'amount': Decimal('10.00000000'), 'confirmations': 1, 'spendable': True, 'solvable': True, 'desc': 'tr([26cdffeb/2/1]9978b573defe38f5d849c2a076a6c666302d49c030ff160ebf40f7ee0a2662dc,{{and_v(v:pk([7a999231/0/1]2d40be15b6fd07671c9249a1ebe321e50e7f1332840cb560324613193e067efb),older(10)),and_v(v:pk([92d0b6b5/0/1]aa6d417376de11e860d2dc78280c6035ac547aa74f09514b4af630242efc7e37),older(10))},and_v(v:pk([0b835bd9/0/1]d30d1e36b2bcd571a07c40342d39e2afd498ac3c914ee50698bdf3df6b5f605f),older(10))})#7asktx0z', 'parent_descs': ['tr(musig([0f056943/86h/1h/0h]tpubDCeEX49avtiXrBTv3JWTtco99Ka499jXdZHBRtm7va2gkMAui11ctZjqNAT9dLVNaEozt2C1kfTM88cnvZCXsWLJN2p4viGvsyGjtKVV7A1,[fdc7b3da/44h/1h/0h]tpubDDdbJUqUksL8pDjkh18Z6YC3kvBW7zufRKnawvvToRh3oTYVvuV3mNyEgav7VZf7JVMvSUxDxAc1MipVETVe6jmTzcaTNMRGf3Gh3eCQFsY,[44fbc2f9/44h/1h/0h]tpubDDMKx1DX6V3wfubHm2jSX11ToHMdYYwmtS3Sm3VJ5ky5efECGPa1kN3zN8hqdeEqrbqez33f8YHLARxUoEGkr2r68ViwXWAxhVcU8Tw1pSV)/2/*,{and_v(v:pk(musig([0f056943/86h/1h/0h]tpubDCeEX49avtiXrBTv3JWTtco99Ka499jXdZHBRtm7va2gkMAui11ctZjqNAT9dLVNaEozt2C1kfTM88cnvZCXsWLJN2p4viGvsyGjtKVV7A1,[44fbc2f9/44h/1h/0h]tpubDDMKx1DX6V3wfubHm2jSX11ToHMdYYwmtS3Sm3VJ5ky5efECGPa1kN3zN8hqdeEqrbqez33f8YHLARxUoEGkr2r68ViwXWAxhVcU8Tw1pSV)/0/*),older(10)),{and_v(v:pk(musig([0f056943/86h/1h/0h]tpubDCeEX49avtiXrBTv3JWTtco99Ka499jXdZHBRtm7va2gkMAui11ctZjqNAT9dLVNaEozt2C1kfTM88cnvZCXsWLJN2p4viGvsyGjtKVV7A1,[fdc7b3da/44h/1h/0h]tpubDDdbJUqUksL8pDjkh18Z6YC3kvBW7zufRKnawvvToRh3oTYVvuV3mNyEgav7VZf7JVMvSUxDxAc1MipVETVe6jmTzcaTNMRGf3Gh3eCQFsY)/0/*),older(10)),and_v(v:pk(musig([fdc7b3da/44h/1h/0h]tpubDDdbJUqUksL8pDjkh18Z6YC3kvBW7zufRKnawvvToRh3oTYVvuV3mNyEgav7VZf7JVMvSUxDxAc1MipVETVe6jmTzcaTNMRGf3Gh3eCQFsY,[44fbc2f9/44h/1h/0h]tpubDDMKx1DX6V3wfubHm2jSX11ToHMdYYwmtS3Sm3VJ5ky5efECGPa1kN3zN8hqdeEqrbqez33f8YHLARxUoEGkr2r68ViwXWAxhVcU8Tw1pSV)/0/*),older(10))}})#862yfmft'], 'safe': True}]
```

According to my calculations the address is still incorrect even tho recognized as `is_mine`. My app calculated first two external addresses:

0. bcrt1p4e7gsehp8xg87gclgf0fct4ng6sgg0a9ctg4laj5u3wuylw0xcfsyh86xw
1. bcrt1pp4294qkmk43rsmzsan3ckecpv9fktckfkgew2knvnpfh65zl0dqqdaytlp

Core calculated addresses with `getnewaddress`:

0. bcrt1pwvnfa5ypmfvnmcguwllehxqreq9ljrvz58hgeqxpwy9xqfta7r6scptqp0
1. bcrt1pgkt8zf54dgkpf66zt56etr8ksctwczn453ec4m2gqydknstuk7mqg0rrlu

Funnily enough, when I use external descriptor from `listdescriptors` and run `deriveaddresses` it matches what my app calculates, not what `getnewaddress` provides.

achow101 committed:
miniscript: Use a reference to key_exp_index in KeyParser

For key_exp_index to count correctly for miniscript expressions,
KeyParser should hold a reference to it.

achow101 committed:
descriptors: Increment key_exp_index in ParsePubkey(Inner)

Simplifies the callsites for incrementing key_exp_index

scgbckbone commented:
tACK https://github.com/bitcoin/bitcoin/pull/34141/commits/2500adcbe7614d5ab34c7ad3463c5b2cb3674643

achow101 commented on src/script/descriptor.cpp:
```diff
@@ -2083,12 +2054,12 @@ struct KeyParser {
     KeyParser(FlatSigningProvider* out LIFETIMEBOUND, const SigningProvider* in LIFETIMEBOUND,
-              miniscript::MiniscriptContext ctx, uint32_t offset = 0)
-        : m_out(out), m_in(in), m_script_ctx(ctx), m_offset(offset) {}
+              miniscript::MiniscriptContext ctx, uint32_t& key_exp_index)
```
Done

sipa commented on src/script/descriptor.cpp:
```diff
@@ -2174,7 +2175,6 @@ std::vector<std::unique_ptr<DescriptorImpl>> ParseScript(uint32_t& key_exp_index
             error = strprintf("combo(): %s", error);
             return {};
         }
-        ++key_exp_index;
```
I tried to @brunoerg-style revert this line, and none of the tests fail (I tried unit and functional tests). Is there a way to get some assurance that the updating is done right?

achow101 replied:
I added a test case to check that the expression index matches the key count. However, this does require 2 test only functions to be added to `Descriptor`, and I'm not entirely sure that I like this approach. But it did reveal a bug in the counting for `OriginPubkeyProvider`.

achow101 committed:
test: Test that key expression indexes match key count

sipa commented on src/script/descriptor.cpp:
```diff
@@ -1041,6 +1047,32 @@ class DescriptorImpl : public Descriptor
     }
+
+    // NOLINTNEXTLINE(misc-no-recursion)
+    uint32_t GetMaxKeyExpr() const override
```
I'd prefer non-recursive versions here, but given that they're test-only, we can ignore that until it becomes a problem I guess?

sipa replied:
(untested)

```c++
 uint32_t GetMaxKeyExpr() const override
 {
    uint32_t max_key_expr{0};
    std::vector<const DescriptorImpl*> todo = {this};
    while (!todo.empty()) {
        const DescriptorImpl* desc = todo.back();
        todo.pop_back();
        for (const auto& p : desc->m_pubkey_args) {
            max_key_expr = std::max(max_key_expr, p->m_expr_index);
        }
        for (const auto& s : desc->m_subdescriptor_args) todo.push_back(s.get());
    }
    return max_key_expr;
}

size_t GetKeyCount() const override
{
    size_t count{0};
    std::vector<const DescriptorImpl*> todo = {this};
    while (!todo.empty()) {
        const DescriptorImpl* desc = todo.back();
        todo.pop_back();
        for (const auto& p : desc->m_pubkey_args) count += p->GetKeyCount();
        for (const auto& s : desc->m_subdescriptor_args) todo.push_back(s.get());
    }
    return count;
}
```

scgbckbone commented:
re-tACK https://github.com/bitcoin/bitcoin/pull/34141/commits/2d44d4176b1f1d790e71aa496c18f3c8229b2a89

achow101 commented:
> which leads me to suspect that diff maybe not actually have an observable effect?

Indeed. The three lines reverted were for incrementing `key_exp_index` after parsing the key for `pkh()`, `combo()`, and `wpkh()`. All of these are expressions in contexts which do not allow any expressions after these are parsed, so incrementing `key_exp_index` isn't actually doing anything as it is never passed into another `PubkeyProvider`. These lines were in fact redundant in the original code.

sipa reviewed (commented):
ACK 4b53cbd69220c1c786bb23a72c0b26a6f78a38f7

rkrux commented:
Concept ACK 4b53cbd because it allows `musig()` while parsing Miniscript, will review soon.

darosior reviewed (commented):
Doesn't having the repro as the first commit break bisectability?

Other than that, Approach ACK 4b53cbd69220c1c786bb23a72c0b26a6f78a38f7. That makes sense to me but i have not closely reviewed the code.

rkrux reviewed (commented):
ACK 4b53cbd69220c1c786bb23a72c0b26a6f78a38f7

+1 on adding the first commit that adds the unit test helpers and also helps in ensuring the fix is correct later.

I do like pushing down the `key_exp_index` increments within the `ParsePubkeyInner` function instead of the caller `ParseScript` - looks cleaner and more logical.

rkrux commented on src/script/descriptor.h:
```diff
@@ -181,6 +181,12 @@ struct Descriptor {
+    virtual uint32_t GetMaxKeyExpr() const = 0;
+
+    /** Get the number of key expressions in this descriptor. Used only for tests */
+    virtual size_t GetKeyCount() const = 0;
```
> I don't think it makes sense to have a Test class outside of tests that the actual class callers should use is a subclass of. Rather it should be the other way around (which is what I thought you originally proposed),

Yeah, I thought of that earlier but later suggested this approach because it came about to be much shorter than the other one - makes sense to me if we don't prefer `Test` structures to be present outside of the tests.
